src/subcommand.rs

```diff
@@ -124,13 +129,27 @@ impl Subcommand {
                 artifacts.push(Artifact::Example(file));
             }
         }
+        // https://doc.rust-lang.org/cargo/reference/cargo-targets.html#binaries
         if bins {
-            for file in utils::list_rust_files(&root_dir.join("src").join("bin"))? {
-                artifacts.push(Artifact::Root(file));
+            if parsed_manifest.package.map_or(true, |p| p.autobins) {
+                for file in utils::list_rust_files(&root_dir.join("src").join("bin"))? {
+                    artifacts.push(Artifact::Root(file));
+                }
+                // TODO: Only if src/main.rs exists and is a binary?
+                artifacts.push(Artifact::Root(package.clone()));
+            }
+
+            for bin in parsed_manifest.bins {
+                artifacts.push(Artifact::Root(bin.name.clone()));
             }
         }
-        if artifacts.is_empty() {
-            artifacts.push(Artifact::Root(package.clone()));
+        // TODO: Check if lib or bin?
+        if artifacts.is_empty() || lib {
```

Comment:
So this bit is wrong, we should add both the library and all binaries unconditionally if no specific thing was requested (ie. if the user doesn't set any of `--lib`, `--bin`, `--bins`, `--examples`, `--test(s)`, `--bench(es)` etc?).

That is perhaps easier to write when the parsing is converted to `Clap` in #15, shall we land that PR first @dvc94ch?